### DISCUSSION TITLE
Simple job visualisation(aka HWUI from Android)

### DISCUSSION THREAD

**Discussion by byme8** [AvaloniaUI/Avalonia#9132] [General]
I have been playing around with Avalonia, creating custom controls, and analyzing their performance on mobile devices.
Android has the [HWUI](https://developer.android.com/topic/performance/rendering/inspect-gpu-rendering) that allows to see when your frame takes more than 16ms.

It looks like that:
![image](https://user-images.githubusercontent.com/5927686/194721086-478721f5-3753-45c9-80d9-5bef2c792389.png)

I created something similar for Avalonia. It allows you quickly see how much time takes to present the UI. For example, If you want to show a bunch of chips at once, you will see the next picture:

https://user-images.githubusercontent.com/5927686/194721702-f27fb60e-083f-4284-a0f5-f2e4d4fc6938.mp4

The horizontal lines provide you with a reference for the duration. The green horizontal line is 16ms, yellow 32ms, and red 100ms. 
The vertical lines on the bottom show the duration of the jobs. These lines can have different colors. For example, the red color is responsible for the layout calculation. Here we can see that the current implementation blocks the main thread for more than 45ms. What is not ideal. Our animations will look, laggy and overall responsiveness would be bad.

We can make it lazier by Task.Yield-ing, so layout calculation will spread between frames. The results for "improved" implementation will look like that:

https://user-images.githubusercontent.com/5927686/194721913-260344ee-76ca-443e-a44e-b1028cd99bc3.mp4

All jobs took less than 32ms which is much better for responsiveness. Yes, it is a synthetic example, but it is really handy with real apps to monitor the responsiveness.

So, my question do you need it in the main Avalonia repo? I suppose, that the current implementation will need some adjustments to align it with the Avalonia best practices. At the moment, there are [changes](https://github.com/AvaloniaUI/Avalonia/compare/master...byme8:Avalonia:wip/feature/simple-hwui#diff-4faf28f797dddc8beb5c2284f04290ccf6e24d3d4bce8ffd0f75edf1bcba3de9) in Avalonia.Base project to hook up for running jobs and new [control](https://github.com/AvaloniaUI/Avalonia/compare/master...byme8:Avalonia:wip/feature/simple-hwui#diff-fdda71f161627ff5a62d5dcd70ff6ef5e8274ae8ebbdc3f1c2cfcb114bfc7bba) to visualize the jobs. Example of usage you can find [here](https://github.com/AvaloniaUI/Avalonia/compare/master...byme8:Avalonia:wip/feature/simple-hwui#diff-fc419becdebcfb3f9d9c15112ce11157cd15201289639874027a80d7d7d43484).

The full list of changes you can find [here](https://github.com/AvaloniaUI/Avalonia/compare/master...byme8:Avalonia:wip/feature/simple-hwui).

The fully working sample is [here](https://github.com/byme8/Avalonia/tree/wip/feature/simple-hwui) in the MobileSandbox. I decided to commit everything to have a working sample. For a final branch, I will keep only changes inside core projects.

Any thoughts?

**Comment by BAndysc** [comment]
Personally I would use this feature so bad, if it was possible to somehow diagnose what was the reason of the spike, something like Unity Profiler:
<img width="1264" alt="image" src="https://user-images.githubusercontent.com/5689666/197039271-18ff42b9-c97c-4c14-a761-e482b13438ea.png">
Tho it probably uses some Mono magic to measure it.

I know we can attach a profiler like dotTrace by jetBrains, but it is not useful for debugging short spikes now.

**Comment by maxkatz6** [comment]
Personally, I have no good idea how this kind of diagnostics is done in other big frameworks or platforms.
I think ideal solution would be to provide additional metadata to the third-party profilers like dotTrace, so they can filter and visualize depending on job types. If that's even possible.
If there is no good way to integrate with third-party profilers, then we can add support to the DevTools and/or overlay control. Not sure about implementation.

**Comment by wieslawsoltes** [comment]
If you are on macOS you can use xcode instruments

**Comment by wieslawsoltes** [comment]
xcode instruments tracing avalonia app on macOS
![image](https://user-images.githubusercontent.com/2297442/197068608-43ea5f05-1331-47ce-a025-352b0a514f4f.png)
![image](https://user-images.githubusercontent.com/2297442/197068826-06eadf91-6c67-46a2-b142-a02bbac9fd83.png)

**Reply by BAndysc** [reply]
While those graphs look nice, you can't get to know much about the cause of delays, right? (Since xcode has only native code debugger, it is not possible to find which code exactly lags, is it?)

**Reply by wieslawsoltes** [reply]
> While those graphs look nice, you can't get to know much about the cause of delays, right? (Since xcode has only native code debugger, it is not possible to find which code exactly lags, is it?)

I am not sure how much information can get , but definitely you can see a lot of information regarding GPU, which I was mainly using it for in past.

**Comment by wieslawsoltes** [comment]
@BAndysc This is in release mode, if you run full debug mode it might show more info I guess
![image](https://user-images.githubusercontent.com/2297442/197330840-683d4f32-6f10-46fc-8815-dd17b7be75cb.png)

**Reply by maxkatz6** [reply]
Same as Xcode, it doesn't seem to provide more useful information than any other good profiler, like dotMemory. Except it is browser oriented and has rendered frames.
While android/unity HWUI shows time spent specifically per framework-specific jobs (render, layout...).

**Reply by wieslawsoltes** [reply]
I guess we need to add some counters for instrumentation inside framework so tooling for performance analysis can be used and be more useful

**Reply by byme8** [reply]
Yes, I came to the same conclusion. At the moment, the dotnet trace doesn't give enough information about what is happening. I see that styling takes a big chunk of the time on Android platform, but it is not clear what and why exactly happens.

So, my next plans are:
- setup basic counters. As far as I understood, it should be easy
- get counters data inside the dev tools for remote machine/mobile device. It may be tricky, but we will see.
- setup build process that allows to remove counters completely to avoid performance impact in release build. I guess the Fody may help, but it is just an idea.

**Reply by jp2masa** [reply]
> setup build process that allows to remove counters completely to avoid performance impact in release build. I guess the Fody may help, but it is just an idea.

`#ifdef` should be enough I think?

**Reply by byme8** [reply]
If you are building from the source code then yes. But the real apps will use the Nuget package that was compiled long ago. #ifdef will not work there.

**Comment by byme8** [comment]
A small update. I created a tool that injects telemetry inside the assembly via Mono.Cecil. The results are exported into the speedscope format. The whole JSON file takes up to 200Mb, so only a screenshot.
![image](https://user-images.githubusercontent.com/5927686/200129610-2ebbbe0c-9295-4caf-89a5-e211a3a90bda.png)

Currently, it requires marking methods with special attributes inside the Avalonia source code. However, it looks like it is possible to avoid it completely. For example, we can list classes and methods we are interested in and inject those attributes right before the benchmark. It would be much better to do it in such way, because it would allow us to avoid any possible overhead when we don't need it.

**Reply by BAndysc** [reply]
Is it possible to profile the code "per frame"? So that if somthing lags, you would see a spike in a specific frame, then you could select the frame to inspect what caused the lag? Looks good anyway!

**Reply by byme8** [reply]
From what I see, the Avalonia doesn't have a concept of a frame. It has a main loop that goes through jobs. Job can do layout calculation, render, and so on.

**Reply by BAndysc** [reply]
But FPS is somehow measured (in dev tools), probably measuring the UI thread. And also when you perform a long task on the UI thread, when the rendering lasts for more than 16ms, then it lags. And that's what I would love to measure.

**Reply by jp2masa** [reply]
IIRC FPS is measured by ticking a counter on every render job. Anyway, if you're running a long task on the UI thread, it's running on a job. So, if the tool displays all jobs, it should show as a big "slice" before the corresponding render "slice" which lagged.
